src/tribler/core/components/ipv8/eva_protocol.py

```diff
@@ -231,32 +137,87 @@ class TransferType(Enum):
     OUTGOING = auto()
 
 
+@dataclass
+class TransferResult:
+    peer: Peer
+    info: bytes
+    data: bytes
+    nonce: int
+
+
 class Transfer:  # pylint: disable=too-many-instance-attributes
     """The class describes an incoming or an outgoing transfer"""
 
     NONE = -1
 
-    def __init__(self, transfer_type, info_binary, data_binary, nonce):
+    def __init__(self, transfer_type: TransferType, info: bytes, data: bytes, data_size: int, block_count: int,
+                 nonce: int, peer: Peer, protocol, future: Optional[Future] = None, window_size: int = 0,
+                 updated: float = 0):
         self.type = transfer_type
-        self.info_binary = info_binary
-        self.data_binary = data_binary
+        self.info = info
+        self.data = data
+        self.data_size = data_size
+        self.block_count = block_count
         self.block_number = Transfer.NONE
-        self.block_count = 0
-        self.attempt = 0
+        self.future = future
+        self.peer = peer
         self.nonce = nonce
-        self.window_size = 0
+        self.window_size = window_size
+        self.updated = updated
+        self.protocol = protocol
+
+        self.attempt = 0
         self.acknowledgement_number = 0
-        self.updated = time.time()
-        self.released = False
+        self.terminated = False
+
+    def finish(self):
+        result = TransferResult(peer=self.peer, info=self.info, data=self.data, nonce=self.nonce)
+        self.terminate(result=result)
+
+    def terminate(self, result: Optional[TransferResult] = None, exception: Optional[Exception] = None):
+        if self.terminated:
+            return
+
+        logger.debug(f'Terminate. Peer: {self.peer}. Transfer: {self}')
+
+        container = self.protocol.incoming if self.type == TransferType.INCOMING else self.protocol.outgoing
+        container.pop(self.peer, None)
+
+        if result:
+            self._terminate_with_result(result)
+        if exception:
+            self._terminate_with_exception(exception)
+
+        self.info = None
+        self.data = None
+        self.peer = None
+        self.protocol = None
+        self.future = None
+
+        self.terminated = True
 
-    def release(self):
-        self.info_binary = None
-        self.data_binary = None
-        self.released = True
+    def _terminate_with_result(self, result: TransferResult):
+        callbacks = self.protocol.receive_callbacks if self.type == TransferType.INCOMING \
+            else self.protocol.send_complete_callbacks
+
+        for callback in callbacks:
+            asyncio.create_task(callback(result))
+
+        if self.future:
+            self.future.set_result(result)
```

Comment:
**question:** can you explain why you swap the future result's setting and the callback's invocation in the last commit?

I think it is better to set the future's result at the beginning of the function before running the callbacks, as callbacks can theoretically access the future.

Reply:
Both options seem equally correct to me. This swap has been made due to `async` EVA refactoring. After introducing async callbacks tests stop passing because now they should wait until all necessary events are raised.

By doing this swap I keep tests unchanged.

Without swap the new test pattern should be:

```python
class MockCommunity(EVAProtocolMixin, Community):  # pylint: disable=too-many-ancestors
    ...
    def __init__(self, *args, **kwargs):
        ...
        self.data_has_been_sent = asyncio.Event()
        ...


    async def on_send_complete(self, result: TransferResult):
        ...
        self.data_has_been_sent.set()
        
    async def test_one_block_binary(self):
        # In this test we send a single transfer from Alice to Bob.
        # The transfer size is less than and `block_size` and therefore it
        # could be send as a single packet.
        data = (b'test1', b'1234', 42)

        await self.alice.eva.send_binary(self.bob.my_peer, *data)
        assert self.bob.most_recent_received_data == data

        await self.alice.data_has_been_sent.wait()
        assert self.alice.most_recent_sent_data == data
```

But probably this is not a really big change and it doesn't add much complexity.

I've reverted this change back.